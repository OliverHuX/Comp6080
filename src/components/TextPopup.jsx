import React from 'react';
import PropTypes from 'prop-types';
import Button from '@material-ui/core/Button';
import Dialog from '@material-ui/core/Dialog';
import DialogActions from '@material-ui/core/DialogActions';
import DialogContent from '@material-ui/core/DialogContent';
import DialogTitle from '@material-ui/core/DialogTitle';

export const TextButton = ({ buttonText = 'OK', handleOnClick }) => {
  return (
    <Button onClick={handleOnClick} color="primary">
      {buttonText}
    </Button>
  )
}
TextButton.propTypes = {
  handleOnClick: PropTypes.func,
  buttonText: PropTypes.string,
};
export function TextPopup ({ open = false, setOpen, title = 'Notification', handleOnClick, buttonText }) {
  const handleClose = () => setOpen(false);
  return (
      <Dialog open={open} onClose={handleClose} aria-labelledby="form-dialog-title" fullWidth>
        <DialogTitle id="form-dialog-title">{title}</DialogTitle>
        <DialogContent>
        </DialogContent>
        <DialogActions>
          <TextButton buttonText={'Cancel'} handleOnClick={handleClose}/>
          <TextButton buttonText={buttonText} handleOnClick={handleOnClick}/>
        </DialogActions>
      </Dialog>
  );
}
TextPopup.propTypes = {
  open: PropTypes.bool,
  setOpen: PropTypes.func,
  title: PropTypes.string,
  handleOnClick: PropTypes.func,
  buttonText: PropTypes.string,
};
