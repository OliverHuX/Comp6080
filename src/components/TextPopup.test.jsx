import React from 'react';
import { TextButton, TextPopup } from './TextPopup';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

describe('TextButton Component', () => {
  const noop = () => {};
  it('triggers onClick event handler when clicked', () => {
    const onClick = jest.fn();
    shallow(<TextButton handleOnClick={onClick}/>).simulate('click');
    expect(onClick).toHaveBeenCalledTimes(1);
    shallow(<TextButton handleOnClick={onClick}/>).simulate('click');
    shallow(<TextButton handleOnClick={onClick}/>).simulate('click');
    expect(onClick).toHaveBeenCalledTimes(3);
  });
  it('uses given buttonText', () => {
    const buttonText = 'YES';
    const button = shallow(<TextButton handleOnClick={noop} buttonText={buttonText}/>);
    expect(button.text()).toBe(buttonText);
  });
  it('uses fallback buttonText if none provided', () => {
    const button = shallow(<TextButton handleOnClick={noop}/>);
    expect(button.text()).toBe('OK');
  });
  it('renders with minimal props (snapshot)', () => {
    const textButton = renderer.create(<TextButton onClick={noop}/>).toJSON();
    expect(textButton).toMatchSnapshot();
  });
  it('renders with provided button text (snapshot)', () => {
    const textButton = renderer.create(<TextButton onClick={noop} buttonText={'YES'}/>).toJSON();
    expect(textButton).toMatchSnapshot();
  });
});

describe('TextPopup Component', () => {
  const noop = () => {};
  const onClick = jest.fn();
  it('does not display the TextPopup when open is default', () => {
    const textPopup = shallow(<TextPopup setOpen={noop} title={'title'}/>);
    expect(textPopup.props().open).toBe(false);
  });
  it('does not display the TextPopup when prop open is false', () => {
    const textPopup = shallow(<TextPopup open={false} setOpen={noop} title={'title'}/>);
    expect(textPopup.props().open).toBe(false);
  });
  it('opens the TextPopup when prop open is true', () => {
    const textPopup = shallow(<TextPopup open={true} setOpen={noop} title={'title'}/>);
    expect(textPopup.props().open).toBe(true);
  });
  it('uses fallback title if none provided', () => {
    const textPopup = shallow(<TextPopup open={true} setOpen={noop}/>);
    expect(shallow(textPopup.props().children[0]).text()).toBe('Notification');
  });
  it('renders with given title', () => {
    const textPopup = shallow(<TextPopup open={true} setOpen={noop} title={'special title'}/>);
    expect(shallow(textPopup.props().children[0]).text()).toBe('special title');
  });
  it('has two buttons with no buttonText given', () => {
    const textPopup = shallow(<TextPopup open={true} setOpen={noop} title={'title'}/>);
    expect(textPopup.find(TextButton).length).toBe(2);
    const target1 = <TextButton ></TextButton>;
    expect(textPopup.dive().containsMatchingElement(target1)).toBe(true);
    const target2 = <TextButton buttonText={'Cancel'} ></TextButton>;
    expect(textPopup.dive().containsMatchingElement(target2)).toBe(true);
  });
  it('has two buttons when buttonText is given', () => {
    const textPopup = shallow(<TextPopup open={true} setOpen={noop} title={'title'} buttonText={'buttonName'}/>);
    expect(textPopup.find(TextButton).length).toBe(2);
    const target1 = <TextButton buttonText={'buttonName'}></TextButton>;
    expect(textPopup.dive().containsMatchingElement(target1)).toBe(true);
    const target2 = <TextButton buttonText={'Cancel'}></TextButton>;
    expect(textPopup.dive().containsMatchingElement(target2)).toBe(true);
  });
  it('triggers onClick event handler when button clicked ', () => {
    const styledHeader = shallow(<TextPopup open={true} setOpen={noop} title={'title'} buttonText={'buttonName'} handleOnClick={onClick} />);
    const button = styledHeader.find(TextButton).last();
    expect(button.length).toBe(1);
    button.props().handleOnClick();
    expect(onClick).toHaveBeenCalledTimes(1);
  })
  it('renders with minimal props (snapshot)', () => {
    const textPopup = renderer.create(<TextPopup setOpen={noop}/>).toJSON();
    expect(textPopup).toMatchSnapshot();
  });
  it('renders with provided title (snapshot)', () => {
    const textPopup = renderer.create(<TextPopup setOpen={noop} title={'special title'}/>).toJSON();
    expect(textPopup).toMatchSnapshot();
  });
  it('renders with provided title and buttonText (snapshot)', () => {
    const textPopup = renderer.create(<TextPopup setOpen={noop} title={'special title'} buttonText={'buttonName'}/>).toJSON();
    expect(textPopup).toMatchSnapshot();
  });
  it('renders with provided title, buttonText and onclick function (snapshot)', () => {
    const textPopup = renderer.create(<TextPopup setOpen={noop} title={'special title'} buttonText={'buttonName'} handleOnClick={onClick}/>).toJSON();
    expect(textPopup).toMatchSnapshot();
  });
});
