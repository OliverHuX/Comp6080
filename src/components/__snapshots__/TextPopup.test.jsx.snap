// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextButton Component renders with minimal props (snapshot) 1`] = `
<button
  className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary"
  disabled={false}
  onBlur={[Function]}
  onDragLeave={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex={0}
  type="button"
>
  <span
    className="MuiButton-label"
  >
    OK
  </span>
</button>
`;

exports[`TextButton Component renders with provided button text (snapshot) 1`] = `
<button
  className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary"
  disabled={false}
  onBlur={[Function]}
  onDragLeave={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex={0}
  type="button"
>
  <span
    className="MuiButton-label"
  >
    YES
  </span>
  <span
    className="MuiTouchRipple-root"
  />
</button>
`;

exports[`TextPopup Component renders with minimal props (snapshot) 1`] = `null`;

exports[`TextPopup Component renders with provided title (snapshot) 1`] = `null`;

exports[`TextPopup Component renders with provided title and buttonText (snapshot) 1`] = `null`;

exports[`TextPopup Component renders with provided title, buttonText and onclick function (snapshot) 1`] = `null`;
